csds-client/client/v3/client.go

```diff
@@ -0,0 +1,390 @@
+// Package client/v3 implements the client interface for v3 transport api version in specific
+package client
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/json"
+	"envoy-tools/csds-client/client"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"time"
+
+	csdspb_v3 "github.com/envoyproxy/go-control-plane/envoy/service/status/v3"
+	envoy_type_matcher_v3 "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+	"github.com/ghodss/yaml"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+)
+
+// ClientV3 implements the interface Client
+type ClientV3 struct {
+	clientConn *grpc.ClientConn
+	csdsClient csdspb_v3.ClientStatusDiscoveryServiceClient
+
+	nodeMatcher []*envoy_type_matcher_v3.NodeMatcher
+	metadata    metadata.MD
+	opts        client.ClientOptions
+}
+
+// parseNodeMatcher parses the csds request yaml from -request_file and -request_yaml to nodematcher
+// if -request_file and -request_yaml are both set, the values in this yaml string will override and
+// merge with the request loaded from -request_file
+func (c *ClientV3) parseNodeMatcher() error {
+	if c.opts.RequestFile == "" && c.opts.RequestYaml == "" {
+		return errors.New("missing request yaml")
+	}
+
+	var nodematchers []*envoy_type_matcher_v3.NodeMatcher
+	if err := parseYaml(c.opts.RequestFile, c.opts.RequestYaml, &nodematchers); err != nil {
+		return err
+	}
+
+	c.nodeMatcher = nodematchers
+
+	// check if required fields exist in nodematcher
+	switch c.opts.Platform {
+	case "gcp":
+		keys := []string{"TRAFFICDIRECTOR_GCP_PROJECT_NUMBER", "TRAFFICDIRECTOR_NETWORK_NAME"}
+		for _, key := range keys {
+			if value := getValueByKeyFromNodeMatcher(c.nodeMatcher, key); value == "" {
+				return fmt.Errorf("missing field %v in NodeMatcher", key)
+			}
+		}
+	default:
+		return fmt.Errorf("%s platform is not supported, list of supported platforms: gcp", c.opts.Platform)
+	}
+
+	return nil
+}
+
+// connWithAuth connects to uri with authentication
+func (c *ClientV3) connWithAuth() error {
```

Comment:
What's v3 specific in this function? Could it be made generic?

Reply:
For auto authentication mode, we need to parse project number from NodeMatcher and add it to `c.metadata` which is version specific. However, I moved `ConnWithJwt` and `ConnWithAutoGcp` to clientUtil to make these codes generic.

Reply:
In this case why not to make it a parameter, parse NodeMatcher outside of generic code (in the version-specific part) and pass parsed value to the common code?

Reply:
I've just modified `ConnToGCPWithAuto()` to 
`func ConnToGCPWithAuto(uri string, projectNum string) (*grpc.ClientConn, metadata.MD, error)`.

In this way, the projectNumber would be parsed from NodeMatcher in the version-specific part and then passed to it to generate metadata.

Reply:
What's the purpose of passing projectNum to ConnToGCPWithAuto, so that it just populates it in metadata and returns back? This seems to be out of scope of what ConnToGCPWithAuto should do, make interface of the function more complicated without obvious reason.

What is the downside of populating metadata before or after ConnToGCPWithAuto() is called?

Reply:
You are right. I think it would be better to populate metadata before calling ConnToGCPWithAuto(). I've just modified it to make ConnToGCPWithAuto() more straightforward.